### PR TITLE
Implement watchCache structure.

### PR DESCRIPTION
**Pull request by wojtek-t**
Ref #10475

This is implementation of what is described as "FIFO cache" in the proposal doc:
https://github.com/GoogleCloudPlatform/kubernetes/blob/master/docs/proposals/apiserver_watch.md

The proof-of-concept of the whole solution is in #12205 (this shows that this API matches our requirements).

@davidopp @nikhiljindal @bprashanth @lavalamp @fgrzadkowski @gmarek 
@smarterclayton @derekwaynecarr @timothysc

before his vacation @lavalamp suggested @nikhiljindal @bprashanth or @davidopp as reviewers

### CODE REVIEW

**Comment by timothysc** [comment]
cc @rrati @jayunit100

File: pkg/client/cache/reflector.go [hunk 1]
```diff
@@ -232,12 +233,20 @@ func (r *Reflector) listAndWatch(stopCh <-chan struct{}) {
...
 
 // syncWith replaces the store's items with the given list.
-func (r *Reflector) syncWith(items []runtime.Object) error {
+func (r *Reflector) syncWith(items []runtime.Object, resourceVersion string) error {
 	found := make([]interface{}, 0, len(items))
 	for _, item := range items {
 		found = append(found, item)
 	}
 
+	myStore, ok := r.store.(*WatchCache)
```
Comment (smarterclayton) [inline pkg/client/cache/reflector.go]: It's weird to do the string->int conversion here.  Would suggest this become an interface VersionedStore that implements ReplaceWithVersion(string).  Reflector shouldn't have to know about int resource versions.

Comment (wojtek-t) [inline pkg/client/cache/reflector.go]: SGTM - will fix that tomorrow.

File: pkg/client/cache/watch_cache.go [hunk 2]
```diff
@@ -0,0 +1,255 @@
...
+	defer w.Unlock()
+	if w.onEvent != nil {
+		w.onEvent(event)
+	}
+	w.updateCache(resourceVersion, event)
+	w.resourceVersion = resourceVersion
+	return w.store.Add(object)
+}
+
+func (w *WatchCache) Update(obj interface{}) error {
```
Comment (davidopp) [inline pkg/client/cache/watch_cache.go]: It looks like a lot of the code in Add, Update, and Delete is duplicated. I'd suggest factoring that code out into a separate function. I think it's fine to leave that for a subsequent PR, though.

File: pkg/client/cache/watch_cache.go [hunk 3]
```diff
@@ -0,0 +1,255 @@
...
+	defer w.Unlock()
+	if w.onEvent != nil {
+		w.onEvent(event)
+	}
+	w.updateCache(resourceVersion, event)
+	w.resourceVersion = resourceVersion
+	return w.store.Delete(object)
+}
+
+// Assumes that lock is already hold for write.
```
Comment (davidopp) [inline pkg/client/cache/watch_cache.go]: hold->held

Comment (wojtek-t) [inline pkg/client/cache/watch_cache.go]: Done.

File: pkg/client/cache/watch_cache.go (hunk 2 above)
Comment (wojtek-t) [inline pkg/client/cache/watch_cache.go]: Done.

File: pkg/client/cache/reflector.go (hunk 1 above)
Comment (wojtek-t) [inline pkg/client/cache/reflector.go]: Done.

**Comment by wojtek-t** [comment]
@davidopp - I addressed the comments.

PTAL

**Comment by davidopp** [comment]
LGTM

File: pkg/client/cache/watch_cache.go [hunk 4]
```diff
@@ -0,0 +1,245 @@
...
+	w.onReplace = onReplace
+}
+
+func (w *WatchCache) SetOnEvent(onEvent func(watch.Event)) {
+	w.Lock()
+	defer w.Unlock()
+	w.onEvent = onEvent
+}
+
+func (w *WatchCache) GetAllEventsSince(resourceVersion uint64) ([]watch.Event, error) {
```
Comment (lavalamp) [inline pkg/client/cache/watch_cache.go]: Late to the party, maybe done already-- can you document the behavior if you ask for e.g. 4 and 4 & 5 are in the list? Do you get both 4 and 5, or just 4? This behavior has to match etcd's behavior.

Comment (wojtek-t) [inline pkg/client/cache/watch_cache.go]: Yes - you will get both 4 and 5 - it doesn't change the currently existing behavior. I can add a comment to it tomorrow.

**Pull request merged**